File: frontend/src/services/generationMode.ts
export type GenerationMode = 'economy' | 'strict' | 'free'
export type FreeProvider = 'deepseek' | 'chatgpt'

export interface GenerationOptions {
  generation_mode: GenerationMode
  economy_mode: boolean
  free_mode: boolean
  free_provider: FreeProvider
}

/** Read the mode at the exact moment an AI job is submitted. */
export const getGenerationOptions = (): GenerationOptions => {
  const stored = localStorage.getItem('mintext:generationMode')
  const generationMode: GenerationMode =
    stored === 'strict' || stored === 'free' ? stored : 'economy'
  const provider = localStorage.getItem('mintext:freeProvider')
  const freeProvider: FreeProvider = provider === 'chatgpt' ? 'chatgpt' : 'deepseek'
  return {
    generation_mode: generationMode,
    economy_mode: generationMode !== 'strict',
    free_mode: generationMode === 'free',
    free_provider: freeProvider,
  }
}

File: backend/app/schemas/generation_mode.py
from typing import Literal

from pydantic import BaseModel, model_validator


GenerationMode = Literal["economy", "strict", "free"]


class GenerationModeOptions(BaseModel):
    """Common mode snapshot attached to every AI job.

    The optional explicit name is used by current desktop clients.  The two
    booleans remain for compatibility with previously released installers.
    """

    generation_mode: GenerationMode | None = None
    economy_mode: bool = False
    free_mode: bool = False
    free_provider: Literal["deepseek", "chatgpt"] = "deepseek"

    @model_validator(mode="after")
    def synchronize_legacy_flags(self):
        if self.generation_mode == "free":
            self.free_mode = True
            self.economy_mode = True
        elif self.generation_mode == "strict":
            self.free_mode = False
            self.economy_mode = False
        elif self.generation_mode == "economy":
            self.free_mode = False
            self.economy_mode = True
        return self
